docs/data/data-grid/column-menu/column-menu.md

```diff
@@ -0,0 +1,164 @@
+
+```tsx
+function CustomColumnMenu(props: GridColumnMenuProps) {
+  return (
+    <GridColumnMenu
+      {...props}
+      components={{
+        // Override `ColumnMenuFilterItem` component
+        ColumnMenuFilterItem: CustomFilterItem,
+      }}
+    />
+  );
+}
+```
+
+### Hiding a menu item
+
+To hide a menu item, you can set its respective slot in `components` to `null`.
+Check [this table](/x/react-data-grid/column-menu/#column-menu-items-and-their-configurations) to know the slot name for each menu item.
+
+```tsx
+function CustomColumnMenu(props: GridColumnMenuProps) {
+  return (
+    <GridColumnMenu
+      {...props}
+      components={{
+        // Hide `ColumnMenuColumnsItem`
+        ColumnMenuColumnsItem: null,
+      }}
+    />
+  );
+}
+```
+
+### Reordering menu items
+
+Every item is assigned a `displayOrder` based on which it is shown before or after other items. It works in ascending order; the smaller the number is, the earlier it is displayed on the list. For new items default value for `displayOrder` is **100**.
+
+You can override `displayOrder` for the default or new items in `componentsProps`.
+
+Check [this table](/x/react-data-grid/column-menu/#column-menu-items-and-their-configurations) to see default `displayOrder` for each menu item.
+
+```tsx
+function CustomColumnMenu(props: GridColumnMenuProps) {
+  return (
+    <GridColumnMenu
+      {...props}
+      componentsProps={{
+        // Swap positions of filter and sort items
+        columnMenuFilterItem: {
+          displayOrder: 0, // Previously `10`
+        },
+        columnMenuSortItem: {
+          displayOrder: 10, // Previously `0`
+        },
+      }}
+    />
+  );
+}
+```
+
+The following demo implements adding, overriding, hiding, and re-ordering of column menu items.
+
+{{"demo": "ReuseColumnMenuGrid.js", "bg": "inline"}}
+
+### Column menu items and their configurations
+
+As a reference, here are the column menu `components` along with the default item components and `displayOrder`.
+
+{{"demo": "ColumnMenuGridReferences.js", "bg": "inline", "hideToolbar": true}}
+
+## Custom menu component
+
+You can also customize and replace the column menu by [passing a fully custom component](/x/react-data-grid/components/#overriding-components) to the `ColumnMenu` slot of the Data Grid. If you want to add some of the default menu items to your custom component, you can import and re-use them.
+
+{{"demo": "CustomColumnMenuGrid.js", "bg": "inline"}}
+
+:::info
+<strong>Tip</strong>: In the `ColumnMenu` component and its items, you receive the prop `colDef` corresponding to the current column; you can use this to conditionally render some items or change some logic.
+:::
+
+## Customize Menu List props
+
+Column menu internally uses [`<MenuList />`](https://mui.com/material-ui/api/menu-list/) component, use `MenuListProps` component prop to update some props for it. Following example sets the menu layout to `dense`.
```

Comment:
I noticed that it doesn't make the menu denser ([by cutting off vertical margins](https://m2.material.io/design/layout/applying-density.html#usage)) but makes it feel smaller in terms of font size. Maybe we should use the same strategy and nomenclature used in [row density](https://mui.com/x/react-data-grid/accessibility/#density). 


I'm also finding it a bit unnecessary to have the 'Props' suffix at `MenuListProps`. Any particular reason not to call this slot `MenuList`?

Reply:
> I noticed that it doesn't make the menu denser ([by cutting off vertical margins](https://m2.material.io/design/layout/applying-density.html#usage)) but makes it feel smaller in terms of font size. Maybe we should use the same strategy and nomenclature used in [row density](https://mui.com/x/react-data-grid/accessibility/#density).

Is this something we can workaround in our code, or should we consider changing the MenuList (in MUI Core)?

Reply:
> I noticed that it doesn't make the menu denser ([by cutting off vertical margins](https://m2.material.io/design/layout/applying-density.html#usage)) but makes it feel smaller in terms of font size.

Thanks @joserodolfofreitas for providing material specs for density, though I feel it refers more to grids, forms, buttons, etc; I couldn't find a reference about menus. Contrarily, [dense menu specs](https://m2.material.io/components/menus#specs) mention setting 8dp top and bottom margin for dividers (both for dense and normal), and that's how [`<MenuList />`](https://mui.com/material-ui/api/menu-list/#main-content) works too.

> Is this something we can workaround in our code, or should we consider changing the MenuList (in MUI Core)?
Maybe we should use the same strategy and nomenclature used in [row density](https://mui.com/x/react-data-grid/accessibility/#density).

Yes technically it's possible by altering margins for dividers internally based on **dense** prop. But keeping in mind menu specification I am not sure if that is the best way to proceed.
Though I personally share your pain point i.e dense menu doesn't *actually* feel much "dense" 😄

Going with fewer dividers and grouping alone items together as much as possible may solve bit of a problem, something like:
![image](https://user-images.githubusercontent.com/12609561/208040498-505f1aaf-f378-4b1f-bad4-e6361938754d.png)
What do you think?

> I'm also finding it a bit unnecessary to have the 'Props' suffix at MenuListProps. Any particular reason not to call this slot MenuList?

Just wanted to be consistent with how material UI [`<Menu />`](https://mui.com/material-ui/api/menu/) component exposes **MenuList** props. I'm okay with the other way around too.

![image](https://user-images.githubusercontent.com/12609561/208044307-dcfd0327-b13b-4dfd-bf80-eeab9faf3c8f.png)

CC @gerdadesign

Reply:
I think "Hide column" and "Manage columns" can be grouped together! 
Filter, Aggregation, and Grouping seem like they are thematically related as ways to adjust how you're viewing the data, but they're not affecting the same singular function. So it seems they are related in proximity, but not in the same menu group.

For example, this is with column visibility grouped at the bottom. It also groups those 3 ways to manipulate the data together. The visibility seems fine because it has a clear category for two options that affect the same action. The 3 data viewing changes feel off to me because they are compared to menu groups like Pinning or Sorting that have a clear singular action with menu items that affect the same action differently. Another way to think about this is if we had cascading menus, Pinning would be one and Sorting would be one, but these 3 would still remain separate. 
<img width="324" alt="Different grouped items 1" src="https://user-images.githubusercontent.com/106991024/208172541-a8011965-5f47-4494-81da-70168c312449.png">

So I would suggest to still keep them close in proximity, but because visibility is one aspect, it makes sense to me to group that together at the bottom! 
<img width="324" alt="Different grouped items 2" src="https://user-images.githubusercontent.com/106991024/208172549-751835b1-00db-40e7-b0f6-07a759fc255f.png">

Reply:
> Contrarily, [dense menu specs](https://m2.material.io/components/menus#specs) mention setting 8dp top and bottom margin for dividers (both for dense and normal), and that's how [<MenuList />](https://mui.com/material-ui/api/menu-list/#main-content) works too.

Right, @MBilalShafi!
I was thinking about the divider margins, but in reality, it seems it's related each item padding. As a reference, I'm looking at [Angular's implementation](https://angularjsmaterial.dev/latest/demo/menu), which directly refers to [material design v1](https://m1.material.io/components/menus.html#menus-specs), but I don't think there's a clear statement that something regarding density has changed on the next versions. I believe that's the path we should pursue as well. 
Making the menu overall smaller by decreasing font sizes can impact the readability.

But we should do it on a different PR (and [issue](https://github.com/mui/mui-x/issues/7265)) so we don't keep increasing the scope of this one.